### PR TITLE
Fix(Core/CFBG): Balance algorithm

### LINKED ISSUE #78: CFBG balance
The level of the players inside a battleground **is not always balanced.**

### Scenario
4 characters
- lvl 10
- lvl 11
- 2x lvl 19

The first character lvl 10 join BG, one lvl 19 join BG -> result **lvl 10 vs lvl 19**

The other two join but the lvl 11 click on "Enter battle" for first -> in this case the CFBG see that the weak faction is the lvl 10 one, so it puts the lvl 11 with the lvl 10 and **later** it puts the lvl 19 with the other lvl 19

result -> 2x lvl 19 vs lvl 10,11

I reproduced this using the conf EvenTeams and LowLevelBalance enabled.

### PR DESCRIPTION
**Pull request by Nefertumm**
Use an algorithm based on averages. 
Closes #78

### CODE REVIEW

File: src/CFBG.cpp [hunk 1]
```diff
@@ -191,19 +191,51 @@ TeamId CFBG::SelectBgTeam(Battleground* bg, Player *player)
...
-                if (
-                    (playerLevel < averagePlayersLevelQueue || // 1
-                    (playerLevel == averagePlayersLevelQueue && player->GetAverageItemLevel() < averagePlayersItemLevelQueue)) // 2
-                    && !balancedClass // check if the team has been balanced already by the class balance logic
-                )
+                if (!balancedClass)
                 {
-                    team = team == TEAM_ALLIANCE ? TEAM_HORDE : TEAM_ALLIANCE; // swap the team
+                    uint32 playerCountH = bg->GetPlayersCountByTeam(TEAM_HORDE);
+                    uint32 playerCountA = bg->GetPlayersCountByTeam(TEAM_ALLIANCE);
```
Comment (Helias) [inline src/CFBG.cpp]:
```suggestion
                    auto playerCountH = bg->GetPlayersCountByTeam(TEAM_HORDE);
                    auto playerCountA = bg->GetPlayersCountByTeam(TEAM_ALLIANCE);
```

**Comment by Helias** [comment]
Reading the code I don't see where you handled the case when:
- player X is lvl 19
- player Y is lvl 10

- Y and X join BG
- Y click on "enter battle", he will be in the weak or strong team?
- X click on "enter battle" after Y

who will be in the weak team? and why?
the teams are balanced?

If Y will be in the strong team and X in the weak one, consider this other case:
- player Z is 10
- player K is 15

- K click on enter battle
- Z click on enter battle after K

I expect that K goes to the weak team and Z in the strong one, but consider that Z and K decrease the current avgLevel of both factions, how your algorithm will behave in this scenario?

In your algorithm you are considering the avg level of the team, but you are not predicting who is joining, you are only consider the "current player" that is joining but not the next-one that is in queue

**Comment by Nefertumm** [comment]
Regarding your first question: The players are the first to enter or after the hypothetical situation on the issue? In the hypothetical situation, the lvl 19 would go on the weak team and the teams would be balanced, because the algorithm tries to balance to the weaker team instead of the stronger team. 
On the second part, this is how the algorithm would be behave, which is wrong because I don't check for people in queue also. 
image
In order to fix this current issue, I will need to check the queue too, you're right :)

I'll update this after monday hopefully, I have finals coming.

**Comment by Nefertumm** [comment]
While I was resting I was thinking about this problem, I tested the results with what you pointed and it worked. The logic can be improved so it wouldn't be so nested. Feel free to do it if you want, if not I'll do it after Monday.

**Comment by Helias** [comment]
I'll review the code and make some tests in these days, thanks!

**Comment by Helias** [comment]
Sorry, I could not test this :(.
Did you test the scenario that I wrote?

**Review by Helias — APPROVED** [review]

**Comment by Annamaria-CC** [comment]
Whats the status on this?

**Comment by Nefertumm** [comment]
Only in-game testing needed for different scenarios like Helias pointed. I'll come back tomorrow from my holidays. I ran before a couple tests but it's a pretty big change to test alone due you'd have to test with a lot of chars simultaneously while having certain conditions.

**Comment by Annamaria-CC** [comment]
hmm whats the worst that can happen?

**Comment by Nefertumm** [comment]
Well, the worst that can happen can be unbalance :P

**Pull request merged**